### PR TITLE
Fix optional properties for pre-material systems

### PR DESCRIPTION
**Pull request by dschwen**
This should address @hugary1995 's case. 

@loganharbour I'm reworking my original PR here. If you want that templated static warehouse thingie you can add that on later (or let me cherry pick a commit)

Refs #18999

### CODE REVIEW

File: framework/src/materials/MaterialPropertyInterface.C [hunk 1]
```diff
@@ -317,3 +317,19 @@ MaterialPropertyInterface::checkExecutionStage()
     mooseError("Material properties must be retrieved during object construction. This is a code "
                "problem.");
 }
+
+const std::vector<MaterialPropertyInterface *>
+MaterialPropertyInterface::getInterfaceObjects(MooseApp & app)
```
Comment (loganharbour) [inline framework/src/materials/MaterialPropertyInterface.C:322]: This was part of my initial design, but once you have one object inheriting from several interfaces that do something like this it gets ugly real quick

Comment (dschwen) [reply framework/src/materials/MaterialPropertyInterface.C:328]:
```
MaterialPropertyInterface::getMaterialPropertyInterfaceObjects
```

:-)

Comment (dschwen) [reply framework/src/materials/MaterialPropertyInterface.C:328]: But, yeah, I'm sure that can be refined once we get to that point

**Comment by dschwen** [comment]
This almost works (and I have added a few more tests), but I'm seeing a curious exodiff that I need to investigate further.

**Comment by dschwen** [comment]
Looks like an out of order material evaluation, i.e. the consuming material is executed before the providing material, leading to an out of date property value (from the previous element)

**Comment by dschwen** [comment]
Current hypothesis is that the material property resolution needs to call `Material::getMaterialProperty` for properties gotten in `Material` objects (rather than the `MaterialPropertyInterface::getMaterialProperty` method, which does not set the dependency info).

**Comment by dschwen** [comment]
The failing stochastic tools tests have multiapps that get created and destroyed. At some point a new app is created at the memory location of an old app, so we have a `&_app` that we've seen before and there are still (long deleted) objects registered under that map key (the `MooseApp *`).

**Comment by hugary1995** [comment]
Thanks!

**Review by loganharbour — DISMISSED** [review]
Approving content that is not my own

File: framework/include/base/MooseApp.h [hunk 2]
```diff
@@ -1115,6 +1132,20 @@ class MooseApp : public ConsoleStreamInterface,
   std::map<std::pair<std::string, std::string>, void *> _lib_handles;
 
 private:
+  ///@{
+  /// Structs that are used in the _interface_registry
+  struct InterfaceRegistryObjectsBase
+  {
+    ~InterfaceRegistryObjectsBase() {}
```
Comment (dschwen) [inline framework/include/base/MooseApp.h:1139]: This needs a `virtual`

File: framework/include/base/MooseApp.h [hunk 3]
```diff
@@ -1115,6 +1132,20 @@ class MooseApp : public ConsoleStreamInterface,
...
+  };
+
+  template <class T>
+  struct InterfaceRegistryObjects : public InterfaceRegistryObjectsBase
+  {
+    std::vector<T *> _objects;
```
Comment (dschwen) [inline framework/include/base/MooseApp.h:1145]: otherwise this vector does not get destroyed properly (and hence you leak memory when pushing into it)

**Review by dschwen — COMMENTED** [review]
I approve Logan's contributions (but GitHub doesn't let me formally do this). The `const` could be added for clarity, but I leave that up to Logan.

**Review by loganharbour — APPROVED** [review]
Approving content that is not my own

**Review by andrsd — COMMENTED** [review]
Idea looks fine. But it seems that what ever is going from an object into app neglects the threading model? As in: if you had multiple copies of an object (one per thread), all of them would get the same requested copy from the `MooseApp`? Maybe it is not a problem, IDK. Or maybe I missed something...

**Comment by loganharbour** [comment]
> Idea looks fine. But it seems that what ever is going from an object into app neglects the threading model? As in: if you had multiple copies of an object (one per thread), all of them would get the same requested copy from the MooseApp? Maybe it is not a problem, IDK. Or maybe I missed something...

If we create `n_threads` copies, `n_threads` objects will be created and we'll have `n_threads` entries in the registry. Which is what we want.

**Review by andrsd — APPROVED** [review]

**Pull request merged**